Add Pokemon Center Event Distribution Man to PKMN Crystal Legacy

towersvault opened the pull request:
Hi Craig and TSP,

For some backstory, I had been super interested in Gen II Event Distribution Pokemon, however giving myself these Pokemon directly just felt like a bit too much of a cheat. Getting hardware setup for this as well didn't seem ideal, so I thought the next next best thing was to just write a mod for Pokemon Crystal.

From my research this hasn't been done in any other ROMs (of course I have no issue if anyone else uses/modifies my code). I built out this mod with what I deemed would make sense and to feel as _natural_ as possible. Event Distribution Pokemon information came from Bulbapedia and Pokemon Center New York's Historical Archive.

Lorewise, the idea is that the Event Distribution Man is a man that travelled from a distant land using the Johto rail system and brought a few gifts with. He travels to Johto every Tuesday and gets to Golden Rod in the day. 
Upon first introduction, the NPC will ask if Ecruteak is north of Golden Rod. Providing the correct answer will allow you the ability to receive gifts from the NPC as his way of saying thanks.

Details about the feature:

- Adds an NPC in the Golden Rod Train Station called "Cliff". Of course this can be changed, just thought I'd throw in my own name for shits and giggles.
- The NPC is only visible in Gold Rod Train Station if `EVENT_RESTORED_POWER_TO_KANTO` had been set, the day is `TUESDAY`, and the time is `DAY`.
- Distribution Gifts match as closely and as reasonably possible to the original Event Distributions. The list is available in `distribution_gifts.asm`. 
- All the Pokemon in the Distribution Gifts list have the unique movesets they were given in the original Event Distributions.
- The NPC will gift you a random entry from the Event Distributions list, meaning you might get doubles as time goes on. You can either receive an egg or an actual Pokemon as defined in the list. Pokemon (not eggs) themselves belong to a different OT, meaning if you're too low level they of course won't obey you. 
- You can only get one gift per week.
- Eggs that the NPC gives you have a 25% chance to be shiny. This can of course be modified on a per-list-entry case.
- I haven't added Celebi as a gift in the list because I feel the GS ball event in PKMN Crystal Legacy is worth it enough to rather do to get Celebi.

Of course if you have any other suggestions, ideas, bug reports, please let me know.
I've tested the code quite extensively and I feel quite confident it should be a clean merge (I tried minimizing adding to already existing code), however of course touch wood just in case.

towersvault committed:
Initial prep commit.
Most of the constants data legwork has been done. Only the actual Distribution Gifts need to be fleshed out still.
After constants data work is done, work can start on the actual logic behind the Distribution Gift Man.

towersvault committed:
Data tables prepped

towersvault committed:
Added Event Distribution Man to Golden Rod Station

towersvault committed:
Typo

towersvault committed:
Changed PCNY_OT_ID to 5 digits

towersvault committed:
Fixed Distrib. Man's time and day appearance
The Event Distribution Man will now only appear on Tuesdays during the Day.

towersvault committed:
Removed Distributions
Distributions wouldn't have really made sense as the Distribution Man will select from the list of gifts at random anyway.

towersvault committed:
Updated some comments and the PCNY OT info.

towersvault committed:
Implemented Pokemon/Egg Giving
Pokemon giving to the player will have the Event Distribution Man's OT,
whereas Eggs will eventually hatch to adopt the player's OT.

towersvault committed:
Made myself the Event Distribution Man.
It's very late and my vapidness have no bounds.

towersvault committed:
Uncommented Event and DateTime checks for Dist. Man

towersvault committed:
Improved resetting EVENT_POKEDISTRIBUTIONMAN_RECEIVEDGIFT
There was a bug where the reset wouldn't occur if the player entered the Golden Rod Train Station at the same time and date a week later.

towersvault committed:
Event Dist. Man leaves station after giving gift

cRz-Shadows commented:
This is a really cool and unique feature! Spoke to some of the team about this and we think it's for sure worth considering for the kanto expansion coming after emerald legacy.

Just a few notes though,
- We think the npc should be locked to post defeating red. Additionally if the plan is to add it in the kanto expansion, the NPC should probably be in kanto rather than johto
- Some of the gift mons are legendaries, which could trivialise the dex completion quest. We think legendaries and mythicals should remain out of the pool until after you have completed the nat dex and done the mew event.
- 25% shiny is a no go, we haven't altered shiny odds anywhere outside of the odd egg, and don't plan to as the normal 1/8192 is synonymous with the older generations, and increasing this makes shinies much less special or interesting to catch or hunt. This comes with an exception if any of these events originally had increased or guarenteed shiny odds, those are fine to give the rate they had in the original event imo

Just a heads up, we potentially have one more update planned for CL prior to the kanto expansion, so I am going to refrain from merging this in until that update is completed.

towersvault commented:
Thank you for checking out the pull request and giving feedback!

So, on the points you mentioned:

1. I think what I was going for is tying the NPC's appearance into the rail system being made to work again, however perhaps it is better to add him in only after Red so there's more postgame. Luckily placing the NPC somewhere else is quite easy, so I'll move him to somewhere else in Kanto, perhaps the train station or port?
2. I was actually a bit skeptical about the addition of the legendaries and mythicals as well to be honest. I was in fact leaning more towards completely removing them from the distribution list depending on what you and the team thought about it. I'll try to have a look at creating two separate data tables for pre and post completion of the National Pokedex and caught Mew.
3. In the Bulbapedia Wiki page it seems I completely missed the shiny odds for the eggs, my mistake. From the Wiki it seems the eggs have a shiny ratio of 15% unless stated otherwise.

No worries on merging after Kanto Expansion. Of course if there's any other changes/features you guys think of, let me know.

towersvault committed:
Dist. Gift egg shiny rates reduced to match real rates
Bulbapedia Wiki mentioned eggs had a 15% shiny rate unless otherwise specified.

towersvault committed:
Moved Dist. Man to Saffron Magnet Train Station
The NPC will now only appear after the player has defeated Red.

towersvault committed:
Split Dist. Gifts into 2 tables based on Dex completion
If the National Dex isn't complete yet, then Legendary and Mythical Pokemon will not be included as possible gifts for the player.

towersvault commented:
Changes made.
I moved the NPC to the Saffron City Train Station, and he'll only appear now after Red had been defeated.
Also, I implemented the pre- and post-National Dex gift selection data table and the logic behind it.

towersvault committed:
Merge branch 'cRz-Shadows:main' into eventdistributionman

ehusby commented:
Any update on getting this great gameplay feature into the next release?

cRz-Shadows commented:
Still planned but the kanto expansion is yet to be started since the current project is FE7 Legacy
